File: comparar_componentes_cr.php
<?php
/**
 * CLI para comparar el campo `components` deserializado de reportes
 * con nombre coincidente entre ambos plugins:
 *   - Propio:    block_configurable_reports
 *   - Alquilado: block_advanced_reports (u otra tabla detectada)
 *
 * Para cada curso indicado:
 *   1. Obtiene reportes de AMBOS plugins.
 *   2. Empareja por nombre (coincidencia exacta y parcial).
 *   3. Deserializa el campo `components` de cada uno con cr_unserialize.
 *   4. Compara recursivamente cada sección del componente.
 *   5. Emite un JSON detallado con las diferencias encontradas.
 *
 * Parámetros:
 *   --config=/ruta/a/config.php   (obligatorio)
 *   --courses=2,3,4               (obligatorio: ids separados por coma)
 *
 * @package   block_configurable_reports
 * @copyright 2026 Awakelab
 * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

$cliargs = [];
foreach (array_slice($argv, 1) as $token) {
    if (preg_match('/^--([^=]+)=(.*)$/s', $token, $m)) {
        $cliargs[$m[1]] = $m[2];
    } else if (preg_match('/^--([^=]+)$/', $token, $m)) {
        $cliargs[$m[1]] = true;
    }
}

function cr_emit(array $payload, int $exitcode): void {
    fwrite(STDOUT, "<<<CR_RESULT>>>" . json_encode($payload, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT) . "<<<END_CR_RESULT>>>\n");
    exit($exitcode);
}

if (empty($cliargs['config']) || !is_readable($cliargs['config'])) {
    cr_emit(['ok' => false, 'fatal' => 'Falta --config con la ruta a config.php válida'], 2);
}
if (empty($cliargs['courses'])) {
    cr_emit(['ok' => false, 'fatal' => 'Falta --courses (ids separados por coma)'], 2);
}

$courseids = array_filter(array_map('intval', explode(',', $cliargs['courses'])));

define('CLI_SCRIPT', true);
define('NO_OUTPUT_BUFFERING', true);
require($cliargs['config']);
global $CFG, $DB;

// Cargar locallib del plugin propio para cr_unserialize.
$ownlocallib = $CFG->dirroot . '/blocks/configurable_reports/locallib.php';
$hasownplugin = is_readable($ownlocallib);
if ($hasownplugin) {
    require_once($ownlocallib);
}

if (!function_exists('cr_unserialize')) {
    cr_emit(['ok' => false, 'fatal' => 'cr_unserialize no disponible. ¿Está instalado block_configurable_reports?'], 2);
}

// ---------------------------------------------------------------------------
// Detectar tabla del plugin alquilado (misma lógica que comparar_reportes_cr.php)
// ---------------------------------------------------------------------------
$dbman = $DB->get_manager();
$rentedtablename = '';

$candidates = ['block_advanced_reports', 'block_adv_reports', 'block_advreports'];
foreach ($candidates as $tname) {
    if ($dbman->table_exists($tname)) {
        $rentedtablename = $tname;
        break;
    }
}

if (empty($rentedtablename)) {
    try {
        $prefix = $CFG->prefix;
        $alltables = $DB->get_records_sql(
            "SELECT table_name FROM information_schema.tables WHERE table_schema = DATABASE() AND table_name LIKE ?",
            ["{$prefix}%adv%report%"]
        );
        if ($alltables) {
            foreach ($alltables as $t) {
                $rentedtablename = str_replace($prefix, '', $t->table_name);
                break;
            }
        }
    } catch (Throwable $e) {
        // Ignorar.
    }
}

if (empty($rentedtablename)) {
    cr_emit(['ok' => false, 'fatal' => 'No se encontró la tabla del plugin alquilado'], 2);
}

// ---------------------------------------------------------------------------
// Funciones de comparación profunda
// ---------------------------------------------------------------------------

/**
 * Deserializa components de forma segura.
 */
function safe_deserialize($raw) {
    if (empty($raw)) {
        return null;
    }
    // Intentar cr_unserialize primero
    $result = cr_unserialize($raw);
    if ($result !== false && $result !== null) {
        return $result;
    }
    // Intentar unserialize nativo
    $result = @unserialize($raw);
    if ($result !== false) {
        return $result;
    }
    // Intentar base64 + unserialize
    $decoded = @base64_decode($raw, true);
    if ($decoded !== false) {
        $result = @unserialize($decoded);
        if ($result !== false) {
            return $result;
        }
    }
    // Intentar json_decode
    $result = @json_decode($raw, true);
    if ($result !== null) {
        return $result;
    }
    return null;
}

/**
 * Normaliza un valor para comparación (convierte objetos a arrays, ordena claves).
 */
function normalize_value($val) {
    if (is_object($val)) {
        $val = (array) $val;
    }
    if (is_array($val)) {
        $normalized = [];
        foreach ($val as $k => $v) {
            $normalized[$k] = normalize_value($v);
        }
        ksort($normalized);
        return $normalized;
    }
    // Normalizar tipos escalares: "0" == 0, etc.
    if (is_numeric($val)) {
        return (string) $val;
    }
    if (is_bool($val)) {
        return $val ? '1' : '0';
    }
    return $val;
}

/**
 * Compara dos estructuras recursivamente y devuelve las diferencias.
 */
function deep_compare($own, $rented, string $path = ''): array {
    $diffs = [];

    $own_n = normalize_value($own);
    $rented_n = normalize_value($rented);

    if ($own_n === $rented_n) {
        return [];
    }

    // Ambos son arrays: comparar clave por clave.
    if (is_array($own_n) && is_array($rented_n)) {
        $allkeys = array_unique(array_merge(array_keys($own_n), array_keys($rented_n)));
        sort($allkeys);
        foreach ($allkeys as $key) {
            $subpath = $path ? "{$path}.{$key}" : (string) $key;
            $own_has = array_key_exists($key, $own_n);
            $rented_has = array_key_exists($key, $rented_n);

            if ($own_has && !$rented_has) {
                $diffs[] = [
                    'path' => $subpath,
                    'type' => 'solo_en_propio',
                    'own_value' => summarize_value($own_n[$key]),
                ];
            } else if (!$own_has && $rented_has) {
                $diffs[] = [
                    'path' => $subpath,
                    'type' => 'solo_en_alquilado',
                    'rented_value' => summarize_value($rented_n[$key]),
                ];
            } else {
                $subdiffs = deep_compare($own_n[$key], $rented_n[$key], $subpath);
                $diffs = array_merge($diffs, $subdiffs);
            }
        }
        return $diffs;
    }

    // Tipos diferentes o valores escalares distintos.
    $diffs[] = [
        'path' => $path ?: '(raíz)',
        'type' => 'valor_diferente',
        'own_value' => summarize_value($own_n),
        'rented_value' => summarize_value($rented_n),
    ];
    return $diffs;
}

/**
 * Resume un valor para el reporte (trunca strings largos).
 */
function summarize_value($val): string {
    if (is_null($val)) {
        return '(null)';
    }
    if (is_bool($val)) {
        return $val ? 'true' : 'false';
    }
    if (is_array($val)) {
        $json = json_encode($val, JSON_UNESCAPED_UNICODE);
        return strlen($json) > 120 ? substr($json, 0, 117) . '...' : $json;
    }
    $str = (string) $val;
    return strlen($str) > 120 ? substr($str, 0, 117) . '...' : $str;
}

/**
 * Intenta emparejar un nombre del propio con uno del alquilado.
 * Primero intenta coincidencia exacta, luego parcial.
 */
function match_report_name(string $own_name, array $rented_names): ?string {
    // Exacta
    if (in_array($own_name, $rented_names, true)) {
        return $own_name;
    }
    // Normalizada (sin espacios extra, lowercase)
    $own_norm = strtolower(trim(preg_replace('/\s+/', ' ', $own_name)));
    foreach ($rented_names as $rn) {
        $rn_norm = strtolower(trim(preg_replace('/\s+/', ' ', $rn)));
        if ($own_norm === $rn_norm) {
            return $rn;
        }
    }
    // Coincidencia parcial: el nombre propio contiene al alquilado o viceversa
    foreach ($rented_names as $rn) {
        $rn_norm = strtolower(trim(preg_replace('/\s+/', ' ', $rn)));
        if (strpos($own_norm, $rn_norm) !== false || strpos($rn_norm, $own_norm) !== false) {
            return $rn;
        }
    }
    return null;
}

// ---------------------------------------------------------------------------
// Comparar por curso
// ---------------------------------------------------------------------------
$output = [];
$summary = [
    'total_courses' => 0,
    'total_pairs' => 0,
    'identical_pairs' => 0,
    'different_pairs' => 0,
    'own_only' => 0,
    'rented_only' => 0,
    'deserialize_errors' => 0,
];

foreach ($courseids as $courseid) {
    $course = $DB->get_record('course', ['id' => $courseid], 'id, fullname, shortname');
    if (!$course) {
        $output[] = ['courseid' => $courseid, 'error' => 'Curso no encontrado'];
        continue;
    }

    $summary['total_courses']++;
    $coursedata = [
        'courseid' => $courseid,
        'fullname' => $course->fullname,
        'shortname' => $course->shortname,
        'pairs' => [],
        'own_only' => [],
        'rented_only' => [],
    ];

    // Obtener reportes del propio.
    $ownreports = [];
    if ($dbman->table_exists('block_configurable_reports')) {
        $ownreports = $DB->get_records('block_configurable_reports', ['courseid' => $courseid], 'id ASC');
    }

    // Obtener reportes del alquilado.
    $rentedreports = [];
    if ($dbman->table_exists($rentedtablename)) {
        // Detectar campo de curso.
        $samplerow = $DB->get_records_sql("SELECT * FROM {{$rentedtablename}} LIMIT 1");
        $columns_info = $samplerow ? array_keys((array) reset($samplerow)) : [];
        $coursefield = in_array('courseid', $columns_info) ? 'courseid' : null;
        if (!$coursefield && in_array('course', $columns_info)) {
            $coursefield = 'course';
        }
        if ($coursefield) {
            $rentedreports = $DB->get_records($rentedtablename, [$coursefield => $courseid], 'id ASC');
        }
    }

    // Indexar por nombre.
    $own_by_name = [];
    foreach ($ownreports as $r) {
        $own_by_name[$r->name] = $r;
    }
    $rented_by_name = [];
    foreach ($rentedreports as $r) {
        $name = $r->name ?? $r->title ?? $r->reportname ?? "id_{$r->id}";
        $rented_by_name[$name] = $r;
    }

    // Emparejar y comparar.
    $matched_rented = [];
    foreach ($own_by_name as $own_name => $own_rep) {
        $rented_match = match_report_name($own_name, array_keys($rented_by_name));

        if ($rented_match === null) {
            $coursedata['own_only'][] = [
                'id' => (int) $own_rep->id,
                'name' => $own_name,
                'type' => $own_rep->type ?? null,
            ];
            $summary['own_only']++;
            continue;
        }

        $matched_rented[] = $rented_match;
        $rented_rep = $rented_by_name[$rented_match];
        $summary['total_pairs']++;

        // Deserializar ambos.
        $own_components = safe_deserialize($own_rep->components);
        $rented_components = safe_deserialize($rented_rep->components);

        $pair = [
            'own_id' => (int) $own_rep->id,
            'rented_id' => (int) $rented_rep->id,
            'own_name' => $own_name,
            'rented_name' => $rented_match,
            'own_type' => $own_rep->type ?? null,
            'rented_type' => $rented_rep->type ?? null,
        ];

        if ($own_components === null && $rented_components === null) {
            $pair['status'] = 'ambos_sin_componentes';
            $pair['identical'] = true;
            $summary['identical_pairs']++;
        } else if ($own_components === null) {
            $pair['status'] = 'error_deserializar_propio';
            $pair['identical'] = false;
            $summary['deserialize_errors']++;
            $summary['different_pairs']++;
        } else if ($rented_components === null) {
            $pair['status'] = 'error_deserializar_alquilado';
            $pair['identical'] = false;
            $summary['deserialize_errors']++;
            $summary['different_pairs']++;
        } else {
            // Comparación profunda.
            $diffs = deep_compare($own_components, $rented_components);
            if (empty($diffs)) {
                $pair['status'] = 'idénticos';
                $pair['identical'] = true;
                $summary['identical_pairs']++;
            } else {
                $pair['status'] = 'diferentes';
                $pair['identical'] = false;
                $pair['diff_count'] = count($diffs);
                $pair['diffs'] = $diffs;
                $summary['different_pairs']++;
            }
        }

        // Comparar también campos de metadatos relevantes.
        $meta_diffs = [];
        if (isset($own_rep->type) && isset($rented_rep->type) && $own_rep->type !== $rented_rep->type) {
            $meta_diffs[] = ['field' => 'type', 'own' => $own_rep->type, 'rented' => $rented_rep->type];
        }
        if (isset($own_rep->pagination) && isset($rented_rep->pagination) && (string)$own_rep->pagination !== (string)$rented_rep->pagination) {
            $meta_diffs[] = ['field' => 'pagination', 'own' => $own_rep->pagination, 'rented' => $rented_rep->pagination];
        }
        if (isset($own_rep->export) && isset($rented_rep->export) && $own_rep->export !== $rented_rep->export) {
            $meta_diffs[] = ['field' => 'export', 'own' => $own_rep->export, 'rented' => $rented_rep->export];
        }
        if (isset($own_rep->jsordering) && isset($rented_rep->jsordering) && (string)$own_rep->jsordering !== (string)$rented_rep->jsordering) {
            $meta_diffs[] = ['field' => 'jsordering', 'own' => $own_rep->jsordering, 'rented' => $rented_rep->jsordering];
        }
        if (!empty($meta_diffs)) {
            $pair['meta_diffs'] = $meta_diffs;
        }

        $coursedata['pairs'][] = $pair;
    }

    // Reportes solo en el alquilado (sin pareja en el propio).
    foreach ($rented_by_name as $rname => $rrep) {
        if (!in_array($rname, $matched_rented, true)) {
            $coursedata['rented_only'][] = [
                'id' => (int) $rrep->id,
                'name' => $rname,
                'type' => $rrep->type ?? null,
            ];
            $summary['rented_only']++;
        }
    }

    $output[] = $coursedata;
}

cr_emit([
    'ok' => true,
    'rented_table' => $rentedtablename,
    'summary' => $summary,
    'courses' => $output,
], 0);

File: comparar_componentes_cr.py
#!/usr/bin/env python3
"""
Compara el campo `components` deserializado de reportes con nombre
coincidente entre ambos plugins:
  - Propio:    block_configurable_reports
  - Alquilado: block_advanced_reports

Sube comparar_componentes_cr.php al servidor, lo ejecuta como www-data,
y genera un reporte detallado de diferencias en consola + archivo JSON.

Usa los cursos de auditoria_resultado.json (los mismos de la migración).
"""

from __future__ import annotations

import json
import posixpath
import sys
import uuid
from pathlib import Path
from typing import Any, Dict, List

from rich.console import Console
from rich.table import Table
from rich.panel import Panel

from cr_common import (
    apply_execution_mode,
    connect_ssh,
    execute_remote,
    is_local_mode,
    load_inventory,
    parse_php_output,
    prompt_execution_mode,
    prompt_server_selection,
    q,
    safe_cleanup,
    safe_text,
)

SCRIPT_DIR = Path(__file__).resolve().parent
INVENTORY_FILE = SCRIPT_DIR / "inventario.json"
AUDIT_SNAPSHOT_FILE = SCRIPT_DIR / "auditoria_resultado.json"
PHP_COMPARATOR = SCRIPT_DIR / "comparar_componentes_cr.php"
OUTPUT_FILE = SCRIPT_DIR / "comparacion_componentes_resultado.json"
REMOTE_TMP_DIR = "/tmp"

console = Console()


def get_audit_courseids() -> List[str]:
    """Lee los courseids afectados de auditoria_resultado.json."""
    if not AUDIT_SNAPSHOT_FILE.exists():
        console.print("[red]No existe auditoria_resultado.json. Ejecuta primero auditar_cr.py.[/red]")
        sys.exit(1)
    snapshot = json.loads(AUDIT_SNAPSHOT_FILE.read_text(encoding="utf-8"))
    courseids = []
    for server_entry in snapshot.get("servers", []):
        for cid in server_entry.get("affected_courseids", []):
            if str(cid) not in courseids:
                courseids.append(str(cid))
    if not courseids:
        console.print("[yellow]La auditoría no detectó cursos afectados.[/yellow]")
        sys.exit(0)
    return courseids


def run_comparison_batch(ssh, remote_php: str, config_path: str, web_user: str,
                         courseids: List[str], sudo_password) -> Dict[str, Any]:
    """Ejecuta el PHP para un lote de cursos y devuelve el resultado."""
    courses_arg = ",".join(courseids)
    command = (
        f"sudo -u {q(web_user)} env HOME=/tmp php {q(remote_php)} "
        f"--config={q(config_path)} --courses={q(courses_arg)}"
    )
    logs = []
    run_log = execute_remote(
        logs, ssh, step="Comparación componentes", command=command,
        sudo_password=sudo_password, timeout=1800, fail_on_error=False,
    )

    payload = parse_php_output(run_log.stdout)
    if payload is None:
        console.print(f"[red]  No se pudo interpretar la salida PHP para cursos {courses_arg[:60]}...[/red]")
        console.print(f"[dim]  exit={run_log.exit_status}[/dim]")
        console.print(f"[dim]  stderr={run_log.stderr[:300]}[/dim]")
        console.print(f"[dim]  stdout={run_log.stdout[:300]}[/dim]")
        return None
    return payload


BATCH_SIZE = 15  # Cursos por lote para evitar exceder memoria PHP


def run_comparison(server: Dict[str, Any], courseids: List[str]) -> Dict[str, Any]:
    """Sube el PHP, lo ejecuta en lotes y fusiona los resultados."""
    moodle_path = str(server["moodle_path"]).rstrip("/")
    config_path = posixpath.join(moodle_path, "config.php")
    web_user = str(server["web_user"])
    sudo_password = server.get("sudo_password") if server.get("sudo_requires_password") else None

    token = uuid.uuid4().hex
    remote_php = posixpath.join(REMOTE_TMP_DIR, f"cr_comp_components_{token}.php")

    ssh = None
    sftp = None
    cleanup_errors: List[str] = []

    try:
        conn_label = "localmente" if is_local_mode(server) else "por SSH"
        console.print(f"[cyan]Conectando {conn_label} a {server['name']}...[/cyan]")
        ssh = connect_ssh(server)
        sftp = ssh.open_sftp()

        # Subir PHP
        sftp.put(str(PHP_COMPARATOR), remote_php)
        sftp.chmod(remote_php, 0o644)

        # Dividir en lotes
        batches = [courseids[i:i + BATCH_SIZE] for i in range(0, len(courseids), BATCH_SIZE)]
        console.print(
            f"[cyan]Comparando componentes para {len(courseids)} cursos "
            f"en {len(batches)} lotes de hasta {BATCH_SIZE}...[/cyan]"
        )

        # Resultado fusionado
        merged = {
            "ok": True,
            "rented_table": None,
            "summary": {
                "total_courses": 0,
                "total_pairs": 0,
                "identical_pairs": 0,
                "different_pairs": 0,
                "own_only": 0,
                "rented_only": 0,
                "deserialize_errors": 0,
            },
            "courses": [],
        }

        for idx, batch in enumerate(batches, 1):
            console.print(f"[cyan]  Lote {idx}/{len(batches)} ({len(batch)} cursos)...[/cyan]")
            payload = run_comparison_batch(
                ssh, remote_php, config_path, web_user, batch, sudo_password,
            )
            if payload is None:
                continue
            if not payload.get("ok"):
                console.print(f"[red]  Lote {idx} error: {payload.get('fatal', '?')}[/red]")
                continue

            # Fusionar
            if payload.get("rented_table"):
                merged["rented_table"] = payload["rented_table"]
            for key in merged["summary"]:
                merged["summary"][key] += payload.get("summary", {}).get(key, 0)
            merged["courses"].extend(payload.get("courses", []))

        return merged

    except Exception as exc:
        console.print(f"[red]Error: {exc}[/red]")
        return {"error": str(exc)}
    finally:
        if ssh is not None:
            safe_cleanup(ssh, f"rm -f {q(remote_php)}", cleanup_errors, sudo_password=sudo_password)
        if sftp is not None:
            try:
                sftp.close()
            except Exception:
                pass
        if ssh is not None:
            try:
                ssh.close()
            except Exception:
                pass


def print_comparison(data: Dict[str, Any]) -> None:
    """Imprime el comparativo detallado en consola."""
    if not data.get("ok"):
        console.print(f"[red]Error: {data.get('fatal', data.get('error', '?'))}[/red]")
        return

    # Resumen global
    summary = data.get("summary", {})
    console.print(Panel(
        f"Cursos analizados: [bold]{summary.get('total_courses', 0)}[/bold]\n"
        f"Pares comparados:  [bold]{summary.get('total_pairs', 0)}[/bold]\n"
        f"Idénticos:         [bold green]{summary.get('identical_pairs', 0)}[/bold green]\n"
        f"Diferentes:        [bold red]{summary.get('different_pairs', 0)}[/bold red]\n"
        f"Solo en propio:    [yellow]{summary.get('own_only', 0)}[/yellow]\n"
        f"Solo en alquilado: [yellow]{summary.get('rented_only', 0)}[/yellow]\n"
        f"Errores deserial.: [dim]{summary.get('deserialize_errors', 0)}[/dim]",
        title="[bold cyan]Resumen de comparación de componentes[/bold cyan]",
        border_style="cyan",
    ))

    for course in data.get("courses", []):
        if course.get("error"):
            console.print(f"\n[red]Curso {course['courseid']}: {course['error']}[/red]")
            continue

        console.print(f"\n[bold cyan]{'='*60}[/bold cyan]")
        console.print(
            f"[bold]Curso {course['courseid']}:[/bold] "
            f"{course.get('fullname', '?')} ({course.get('shortname', '?')})"
        )

        pairs = course.get("pairs", [])
        own_only = course.get("own_only", [])
        rented_only = course.get("rented_only", [])

        if pairs:
            table = Table(title="Reportes emparejados por nombre")
            table.add_column("Propio ID", justify="right")
            table.add_column("Alquilado ID", justify="right")
            table.add_column("Nombre")
            table.add_column("Estado")
            table.add_column("Difs.", justify="right")

            for pair in pairs:
                if pair.get("identical"):
                    estado = "[bold green]IDÉNTICO[/bold green]"
                    difs = "0"
                elif pair.get("status") == "error_deserializar_propio":
                    estado = "[red]error propio[/red]"
                    difs = "—"
                elif pair.get("status") == "error_deserializar_alquilado":
                    estado = "[red]error alquilado[/red]"
                    difs = "—"
                else:
                    estado = "[yellow]DIFERENTE[/yellow]"
                    difs = str(pair.get("diff_count", "?"))

                name_display = pair["own_name"]
                if pair["own_name"] != pair["rented_name"]:
                    name_display += f" ↔ {pair['rented_name']}"

                table.add_row(
                    str(pair["own_id"]), str(pair["rented_id"]),
                    name_display[:50], estado, difs,
                )
            console.print(table)

            # Mostrar detalles de diferencias
            for pair in pairs:
                if pair.get("identical") or not pair.get("diffs"):
                    continue

                console.print(f"\n  [bold yellow]Diferencias en:[/bold yellow] {pair['own_name']}")
                for diff in pair["diffs"][:15]:  # Limitar a 15 diferencias por reporte
                    path = diff.get("path", "?")
                    dtype = diff.get("type", "?")

                    if dtype == "solo_en_propio":
                        console.print(f"    [green]+ {path}[/green]: {diff.get('own_value', '')}")
                    elif dtype == "solo_en_alquilado":
                        console.print(f"    [red]- {path}[/red]: {diff.get('rented_value', '')}")
                    else:
                        console.print(
                            f"    [yellow]~ {path}[/yellow]: "
                            f"propio=[cyan]{diff.get('own_value', '')}[/cyan] | "
                            f"alquilado=[magenta]{diff.get('rented_value', '')}[/magenta]"
                        )

                if len(pair.get("diffs", [])) > 15:
                    console.print(f"    [dim]... y {len(pair['diffs']) - 15} diferencia(s) más[/dim]")

                # Mostrar diferencias de metadatos
                if pair.get("meta_diffs"):
                    for md in pair["meta_diffs"]:
                        console.print(
                            f"    [dim]meta.{md['field']}:[/dim] "
                            f"propio={md['own']} | alquilado={md['rented']}"
                        )

        if own_only:
            console.print(f"\n  [yellow]Solo en propio ({len(own_only)}):[/yellow]")
            for r in own_only:
                console.print(f"    • {r['name']} (id={r['id']}, tipo={r.get('type', '?')})")

        if rented_only:
            console.print(f"\n  [yellow]Solo en alquilado ({len(rented_only)}):[/yellow]")
            for r in rented_only:
                console.print(f"    • {r['name']} (id={r['id']}, tipo={r.get('type', '?')})")


def main() -> None:
    console.print("[bold cyan]Comparador de componentes · Propio vs. Alquilado[/bold cyan]\n")

    try:
        servers, settings = load_inventory(INVENTORY_FILE)
        exec_mode = prompt_execution_mode()
        apply_execution_mode(servers, exec_mode)

        selected_servers = prompt_server_selection(servers)
        if not selected_servers:
            console.print("[yellow]No se seleccionaron plataformas.[/yellow]")
            return

        courseids = get_audit_courseids()
        console.print(f"[green]Cursos a comparar (de la auditoría):[/green] {', '.join(courseids)}")

        server = selected_servers[0]
        result = run_comparison(server, courseids)

        # Guardar resultado completo
        OUTPUT_FILE.write_text(json.dumps(result, indent=2, ensure_ascii=False), encoding="utf-8")
        console.print(f"\n[green]Resultado guardado en:[/green] {OUTPUT_FILE}")

        # Mostrar comparativo en consola
        print_comparison(result)

    except KeyboardInterrupt:
        console.print("\n[yellow]Cancelado.[/yellow]")
    except Exception as exc:
        console.print(f"[bold red]Error fatal: {exc}[/bold red]")
        sys.exit(1)


if __name__ == "__main__":
    main()